mmv1/third_party/terraform/utils/framework_location.go

```diff
@@ -0,0 +1,99 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type LocationDescriber interface {
+	getLocationDescription(providerConfig *frameworkProvider) LocationDescription
+}
+
+type LocationDescription struct {
+	LocationSchemaField types.String
+	RegionSchemaField   types.String
+	ZoneSchemaField     types.String
+
+	ResourceLocation types.String
+	ResourceRegion   types.String
+	ResourceZone     types.String
+
+	ProviderRegion types.String
+	ProviderZone   types.String
+}
+
+func (ld *LocationDescription) getLocation() (types.String, error) {
+	// Location from resource config
+	if !ld.ResourceLocation.IsNull() && !ld.ResourceLocation.IsUnknown() {
+		return ld.ResourceLocation, nil
+	}
+
+	// Location from region in resource config
+	if !ld.ResourceRegion.IsNull() && !ld.ResourceRegion.IsUnknown() {
+		return ld.ResourceRegion, nil
+	}
+
+	// Location from zone in resource config
+	if !ld.ResourceZone.IsNull() && !ld.ResourceZone.IsUnknown() {
+		location := GetResourceNameFromSelfLink(ld.ResourceZone.ValueString()) // Zone could be a self link
+		return types.StringValue(location), nil
+	}
+
+	// Location from zone in provider config
+	if !ld.ProviderZone.IsNull() && !ld.ProviderZone.IsUnknown() {
+		return ld.ProviderZone, nil
+	}
+
+	return types.StringNull(), nil
```

Comment:
should this return an error rather than nil?

Reply:
This was a big thing for me to miss 😓, thanks for spotting it! I've updated the function and updated the tests to fail if an error isn't returned when one is expected in [ef5f631](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/ef5f6314b0cce05006ef92e24c4c00ce5271a960)